package-res/resources/web/pentaho/visual/base/View.js

```diff
@@ -94,37 +91,97 @@ define([
     /**
      * Orchestrates the rendering of the visualization.
      *
-     * This method executes [_render]{@link pentaho.visual.base.View#_render}
+     * Executes the will/did/rejected event loop associated with an update in a visualization.
+     *
+     * This method executes [_doUpdate]{@link pentaho.visual.base.View#_doUpdate}
      * asynchronously and is meant to be invoked by the container.
      *
      * @return {Promise} A promise that is fulfilled when the visualization
      * is completely rendered. If the visualization is in an invalid state, the promise
      * is immediately rejected.
+     *
+     * @fires "will:update"
+     * @fires "did:update"
+     * @fires "rejected:update"
+     * @fires "did:create"
      */
-    render: function() {
-      var me = this;
+    update: function() {
+      var me = this,
+          model = this.model,
+          isFirstRender = !this._element;
+
+      var willUpdate = new WillUpdate(model);
+      if(model._hasListeners(willUpdate.type))
+        model._emitSafe(willUpdate);
+
       return new Promise(function(resolve, reject) {
-        try {
-          var validationErrors = me._validate();
-          if(!validationErrors) {
-            Promise.resolve(me._render()).then(resolve, reject);
-          } else {
-            reject(validationErrors);
-          }
-        } catch(e) {
-          reject(e.message);
+        var result = willUpdate.isCanceled ? ActionResult.reject(willUpdate.cancelReason)
+                                           : me._doUpdate();
+
+        if(result.error) {
+          if(model._hasListeners(RejectedUpdate.type))
+            model._emitSafe(new RejectedUpdate(model, result.error));
+
+          reject(result.error);
+        } else {
+          if(model._hasListeners(DidUpdate.type))
+            model._emitSafe(new DidUpdate(model));
+
+          if(isFirstRender && model._hasListeners(DidCreate.type))
+            model._emitSafe(new DidCreate(model, me._element));
+
+          resolve();
         }
+
       });
     },
 
     /**
+     * Updates a visualization.
+     *
+     * If the visualization is valid, the visualization element will be created on the first update
+     * and proceed with the visualization update, otherwise it will be rejected and prevent the update.
+     *
+     * @return {pentaho.lang.ActionResult} The result object.
+     * @protected
+     */
+    _doUpdate: function() {
+      var result;
+      try {
+        var validationErrors = this._validate();
+        if(!validationErrors) {
+          if(!this._element) this.prepare();
+
+          this._update();
+          result = ActionResult.fulfill();
+
+        } else {
+          var error = Array.isArray(validationErrors) ? validationErrors.join("\n - ") : validationErrors;
```

Comment:
Didn't notice that I forgot to give a header to the error. With that the current implementation will make more sense